src/puppetlabs/puppetdb/query_eng/engine.clj

```diff
@@ -1666,38 +1667,51 @@
   [query-rec
    columns]
   (->> columns
-       (map #(if (= "function" (first %)) (second %) %))
+       (map #(if (function-call? %) (second %) %))
        (sort-by #(if (string? %) % (:column %)))
        (mapv (partial alias-columns query-rec))))
 
-(defn strip-function-calls
-  [column-or-columns]
-  (let [{functions true
-         nonfunctions false} (->> column-or-columns
-                                  utils/vector-maybe
-                                  (group-by (comp #(= "function" %) first)))]
-    [(vec (map rest functions))
-     (vec nonfunctions)]))
+(defn normalize-function-calls
+  [query-rec function-fields]
+  (seq (map (fn [[name & args]]
+              (apply vector
+                     name
+                     (if (empty? args)
+                       [:*]
+                       (map (partial coalesce-fact-values query-rec) args))))
+            function-fields)))
+
+(pls/defn-validated create-from-node
+  :- {(s/optional-key :projected-fields) [projection-schema]
+      s/Any s/Any}
+  "Create an explicit subquery declaration to mimic the select_<entity>
+   syntax."
+  [entity expr clauses]
+  (let [query-rec (subquery->query-rec (str "select_" (utils/dashes->underscores entity)))
+        {:keys [limit offset order-by]} (create-paging-map clauses)]
+    (if (extract-expression? expr)
+      (let [[_ columns & clauses] expr
+            [query-expr groupby] (map first (split-with #(not= "group_by" (first %)) clauses))
```

Comment:
that won't quite work, since this is extracting both query-expr and group-by and accounting for the cases where either might be missing. I can think about cleaner ways to write it though.